apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/ReliableStreamReader.cs

```diff
@@ -0,0 +1,258 @@
+        private readonly SpannerClient _spannerClient;
+        private readonly ExecuteSqlRequest _request;
+        private readonly Session _session;
+        private readonly CallSettings _callTiming;
+        private ResultSetMetadata _metadata;
+        private bool _isReading = true;
+
+        private ByteString _resumeToken;
+        private int _resumeSkipCount;
+
+        internal ReliableStreamReader(SpannerClient spannerClientImpl, ExecuteSqlRequest request, Session session)
+        {
+            _spannerClient = spannerClientImpl;
+            _request = request;
+            _session = session;
+            _clock = SpannerSettings.GetDefault().Clock ?? SystemClock.Instance;
+            _scheduler = SpannerSettings.GetDefault().Scheduler ?? SystemScheduler.Instance;
+
+            _callTiming = CallSettings.FromCallTiming(CallTiming.FromRetry(new RetrySettings(
+                    SpannerSettings.GetDefaultRetryBackoff(),
+                    SpannerSettings.GetDefaultTimeoutBackoff(),
+                    Expiration.FromTimeout(TimeSpan.FromMilliseconds(600000)),
+                    SpannerSettings.IdempotentRetryFilter
+                )));
+            _request.SessionAsSessionName = _session.SessionName;
+        }
+
+        private async Task<Metadata> ConnectAsync()
+        {
+            Logger.LogPerformanceCounterFn("StreamReader.ConnectCount", x => x + 1);
+            if (_resumeToken != null)
+            {
+                Logger.Debug(() => $"Resuming at location:{_resumeToken}");
+                _request.ResumeToken = _resumeToken;
+            }
+            _currentCall = _spannerClient.ExecuteSqlStream(_request);
+            return await _currentCall.ResponseHeadersAsync.ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Connects or reconnects to Spanner, fast forwarding to where we left off based on
+        /// our _resumeToken and _resumeSkipCount.
+        /// </summary>
+        /// <returns></returns>
+        private async Task<bool> ReliableConnect(CancellationToken cancellationToken)
+        {
+            if (_currentCall == null)
+            {
+//                Func<Task<Metadata>> connectFn = ConnectAsync;
+//                var responseHeaders = await connectFn.CallWithRetry(_callTiming, _clock, _scheduler);
+                await ConnectAsync().ConfigureAwait(false);
+
+                Debug.Assert(_currentCall != null, "_currentCall != null");
+                cancellationToken.ThrowIfCancellationRequested();
+
+                for (int i = 0; i <= _resumeSkipCount; i++)
+                {
+                    //This calls a simple movenext on purpose.  If we get an error here, we'll fail out.
+                    _isReading = await _currentCall.ResponseStream.MoveNext(cancellationToken);
+                    if (!_isReading || _currentCall.ResponseStream.Current == null)
+                    {
+                        return false;
```

Comment:
dont think so.  the entire reader should be disposed by the datareader above it.